src/server/repositories/UrlRepository.ts

```diff
@@ -136,6 +141,57 @@ export class UrlRepository implements UrlRepositoryInterface {
     await url.increment('clicks')
   }
 
+  public updateDailyStatistics: (shortUrl: string) => Promise<void> = async (
+    shortUrl,
+  ) => {
+    const time = getLocalTime()
+    const [clickStats] = await Clicks.findOrCreate({
+      where: { shortUrl, date: time.date },
+    })
+    clickStats.increment('clicks')
+  }
+
+  public updateWeekdayStatistics: (shortUrl: string) => Promise<void> = async (
+    shortUrl,
+  ) => {
+    const time = getLocalTime()
+    const [clickStats] = await WeekdayClicks.findOrCreate({
+      where: { shortUrl, weekday: time.weekday, hours: time.hours },
+    })
+    clickStats.increment('clicks')
+  }
+
+  public updateDeviceStatistics: (
+    shortUrl: string,
+    userAgent: string,
+  ) => Promise<void> = async (shortUrl, userAgent) => {
+    const deviceType = getDeviceType(userAgent)
+    if (deviceType) {
+      const [clickStats] = await Devices.findOrCreate({
+        where: { shortUrl },
+      })
+      clickStats.increment(deviceType!)
+    }
+  }
+
+  public updateLinkStatistics: (
+    shortUrl: string,
+    userAgent: string,
+  ) => Promise<void> = async (shortUrl, userAgent) => {
+    const urlExists = Boolean(await Url.findOne({ where: { shortUrl } }))
```

Comment:
is this check really necessary? it will be yet another network call to the DB

Reply:
It was mostly for caution as I wanted to be very sure that nonexistent urls do not get added to the stats tables. In regards to cutting down on network calls, this check can definitely be removed, since the service layer already checks whether the user owns the short url.

Reply:
Alright, I have removed the check